Make tiddler file paths configurable

nome opened the pull request:
When saving new tiddlers on node.js, allow the user to override the path of the
generated .tid file. This is done by creating a tiddler
$:/config/FileSystemPaths which contains one or more filter expressions, one
per line. These filters are applied in turn to the tiddler to be saved, and
the first output produced is taken as a logical path relative to the wiki's
tiddlers directory. Any occurences of "/" in the logical path are replaced with
the platform's path separator, the extension ".tid" is appended, illegal
characters are replaced by "_" and the path is disambiguated (if necessary) in
order to arrive at the final tiddler file path. If none of the filters matches,
or the configuration tiddler does not exist, fall back to the previous file
naming scheme (i.e. replacing "/" by "_").

This implies we will now, for tiddlers matching the user-specified filters,
create directory trees below the tiddlers directory. In order to avoid
cluttering it with empty directory trees when renaming or removing tiddlers, any
directories that become empty by deleting a tiddler file are removed
(recursively).

Benefits of this configuration option include the ability to organize git
repositories of TiddlyWikis running on node.js, ability to replace characters
that cause trouble with particular operating systems or workflows (e.g. '$' on
unix) and the ability to replicate tiddler "paths" in the filesystem (by
including a filter like "[!has[draft.of]]") without forcing such a (potentially
problematic) change on all users.

---

This is a more ambitious version of #2360, so these PRs are mutually exclusive. Tested with node.js 0.10.25 on Linux. Should receive further testing on different configurations to make sure it doesn't break anything.

Example:

```
[is[system]removeprefix[$:/]addprefix[_system/]]
[tag[task][addprefix[mytasks/tiddlers/]]
[!has[draft.of]]
```

The first line causes system tiddlers to be written to `tiddlers/_system/`, minus the `$:/` prefix (as suggested by @sukima and initially implemented in #2360). The second line causes tiddlers tagged `task` to be written to a separate subdirectory, as suggested by @Jermolene. The third line causes
 logical "paths" of all other non-draft tiddlers to be replicated in the filesystem.

sukima commented:
this is impressive. Kudos nome. Do you need help rebasing this since #2360 has already been merged? Or is this work done after?

I think it would gap to have some doc toddlers included with this so it is available on tiddlywiki.com when this merges. Do you want help with that as well?

Jermolene commented:
Hi @nome that's great, thank you. Docs would indeed be useful, and there are some minor coding style issues that I'll comment on overleaf.

Jermolene commented on plugins/tiddlywiki/filesystem/filesystemadaptor.js:
```diff
@@ -88,11 +88,41 @@ Transliterate string from cyrillic russian to latin
 /*
+Given a list of filters, apply every one in turn to source, and return the first result of the first filter with non-empty result.
+*/
+FileSystemAdaptor.prototype.findFirstFilter = function(filters, source) {
```
Our coding style is a little unusual, I'm afraid, and still a little inconsistent. Anyhow, we don't use spaces within parameter lists in function calls or definitions.

Jermolene commented on plugins/tiddlywiki/filesystem/filesystemadaptor.js:
```diff
@@ -88,11 +88,41 @@ Transliterate string from cyrillic russian to latin
+	if (pathNameFilters) {
+		var source = this.wiki.makeTiddlerIterator([title]);
+		var result = this.findFirstFilter(pathNameFilters.split("\n"), source);
+		if (result) {
```
We omit the space between "if" and the opening bracket, and similarly for `for`

Jermolene commented on plugins/tiddlywiki/filesystem/filesystemadaptor.js:
```diff
@@ -162,6 +196,27 @@ FileSystemAdaptor.prototype.loadTiddler = function(title,callback) {
 /*
+Recursively delete a tree of empty directories
+*/
+FileSystemAdaptor.prototype.deleteEmptyDirs = function(dirpath,callback) {
```
I wonder if this should be in `$:/core/modules/utils/filesystem.js`?

nome commented:
@sukima Thanks! I've added some basic docs now, but any improvements would be most welcome. Also, help with testing, and perhaps refining the design (if necessary) would be great.

@Jermolene Done, and done. I've moved `deleteEmptyDirs` to `core/modules/utils/filesystem.js` as you suggested. Should `findFirstFilter` also be moved to the utils namespace, perhaps with a better name?

Jermolene commented:
Thanks @nome 

> Should `findFirstFilter` also be moved to the utils namespace, perhaps with a better name?

I wondered about moving it into `filter.js` but ended up uncertain whether it is actually sufficiently useful and non-trivial. I think let's leave it where it is for the moment.

Jermolene commented:
Thanks @nome 

I haven't had a chance to test it extensively so I'd be very grateful if other Node.js users could give it a spin.
